### PR TITLE
HIVE-27854:Driver compile may causing memory leak

### PR DESCRIPTION
**Pull request by chenruotao**
HIVE-27854:Driver compile may causing memory leak

### What changes were proposed in this pull request?

I describe it in this the issuse 
https://issues.apache.org/jira/browse/HIVE-27854

### Why are the changes needed?

I describe it in this the issuse 
https://issues.apache.org/jira/browse/HIVE-27854

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

I describe it in this the issuse 
https://issues.apache.org/jira/browse/HIVE-27854

### CODE REVIEW

**Comment by chenruotao** [comment]
@miklosgergely hi, could you tell me why my code test fail ? please

**Comment by chenruotao** [comment]
@aturoczy Hi i do test on master, this commit code can stop driver go on compiling after be closed or destroyed, but master would not cause memory leak even no this commit code, anyway, i still think the driver should not go on compiling after that, maybe cause other problems, please take a look.

**Review by aturoczy — APPROVED** [review]
+1
It is follow-up on a different PR
cc: @deniskuzZ @ayushtkn @SourabhBadhya

**Comment by deniskuzZ** [comment]
@chenruotao, code linked in a JIRA is not relevant for Hive 4. Could you reproduce the same issue in master branch?

**Comment by chenruotao** [comment]
> in SQLOperation#cleanup we call driver.close before settting it to NULL
> 
> ```
>     if (driver != null) {
>       driver.close();
>       driver.destroy();
>     }
>     driver = null;
> ```

In my test, set driver = null can not stop the compiling thread, only set aborted = true can interrupt the thread, the thread still run after set null, and it is the root cause of  memory leak in Hive 3

File: ql/src/java/org/apache/hadoop/hive/ql/DriverState.java [hunk 1]
```diff
@@ -89,7 +89,11 @@ public void compiling() {
   public void compilingWithLocking() {
     lock();
     try {
-      driverState = State.COMPILING;
+      if (isDestroyed() || isClosed()) {
+        abort();
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/DriverState.java:93]: `abort` is only used if the query is in the middle of compilation or execution 
````
if (driverState.isCompiling() || driverState.isExecuting()) {
        driverState.abort();
}
````

Comment (chenruotao) [reply ql/src/java/org/apache/hadoop/hive/ql/DriverState.java:93]: yes, but the question is that the driver may be set null when the query is in the middle of compilation, and if the compiling thread just runs before lock(), the driverState will be reset to State.COMPILING from State.CLOSED OR State.DESTROYED, and keep running.

**Pull request closed without merging**